Added check to ignore semicolon in parser

AngelOnFira opened the pull request:
<!---
Thank you for contributing to Boa! Please fill out the template below, and remove or add any
information as you feel neccesary.
--->

This Pull Request closes #892.

It changes the following:

- Empty statements are now parsed without error
- Semicolons are no longer skipped at the end of the parse loop

In the original issue, it was mentioned that there might be a problem with empty `StatementLists` as well. I couldn't find a case where it didn't work as is, but I still brought the semicolon skip up a little higher.

I'm not sure if any specific tests should be written for this, but if so I can add them.

This PR improves conformance by 0.07% :smile:

Lan2u commented on boa/src/syntax/parser/statement/mod.rs:
```diff
@@ -214,22 +214,24 @@ where
     allow_yield: AllowYield,
     allow_await: AllowAwait,
     allow_return: AllowReturn,
     break_when_closingbraces: bool,
```
Looks really good! Just a couple of minor things which might be nice to include. 'break_when_closingbraces' no-longer seems to actually reflect the new behaviour as any of the break nodes will trigger the check - can you either rename it or look to remove it entirely.

Razican replied:
I think that if we use `}` as a `break_node`, then we don't need this anymore, and can remove it from the structure.

AngelOnFira replied:
Ok, I took it out and everything appears to still be fine.

Lan2u reviewed (commented):
Looks really good! Just a couple of minor things which might be nice to include :)

Some tests showing the empty statement parsing etc. would also be nice - if the code is going to change later I'd argue they are even more valuable

croraf commented:
I suggested on Discord to split this into two PRs. This one with a parse method refactor. And the new one with the EmptyStatement correct parsing solution.

For the second one the idea is not to skip the semicolon but to parse it into EmptyStatement node, as does esprima.org parser:
![image](https://user-images.githubusercontent.com/19698636/97161105-9e5a4900-177d-11eb-8e91-37fd9402ee8d.png)

Razican commented:
I will review this as soon as I have time, hopefully today.

> I suggested on Discord to split this into two PRs. This one with a parse method refactor. And the new one with the EmptyStatement correct parsing solution.
> 
> For the second one the idea is not to skip the semicolon but to parse it into EmptyStatement node, as does esprima.org parser

This makes sense from a Parser-only perspective, but the issue with this is that adds extra allocations in the `StatementList`, by adding some `EmptyStatement` nodes that will need to be dynamically called `run()` during the execution, which will be just a no-op, but will increase cache misses and memory usage, note that all node variants have the same size in memory.

The current version just ignores `EmptyStatement`s, and will not add them to an `StatementList`, which is far more efficient. And having an `EmptyStatement` in the `StatementList` doesn't really provide any useful information. The only use case I can think of would be if we are trying to re-generate the original code, but we are already ignoring comments, for example, so this is currently not possible, and I think it should be the application who extracts the original code again if needed.

In my opinion we shouldn't add an `EmptyStatement` parsing and node, for performance reasons.

croraf commented:
> I will review this as soon as I have time, hopefully today.
> 
> > I suggested on Discord to split this into two PRs. This one with a parse method refactor. And the new one with the EmptyStatement correct parsing solution.
> > For the second one the idea is not to skip the semicolon but to parse it into EmptyStatement node, as does esprima.org parser
> 
> This makes sense from a Parser-only perspective, but the issue with this is that adds extra allocations in the `StatementList`, by adding some `EmptyStatement` nodes that will need to be dynamically called `run()` during the execution, which will be just a no-op, but will increase cache misses and memory usage, note that all node variants have the same size in memory.
> 
> The current version just ignores `EmptyStatement`s, and will not add them to an `StatementList`, which is far more efficient. And having an `EmptyStatement` in the `StatementList` doesn't really provide any useful information. The only use case I can think of would be if we are trying to re-generate the original code, but we are already ignoring comments, for example, so this is currently not possible, and I think it should be the application who extracts the original code again if needed.
> 
> In my opinion we shouldn't add an `EmptyStatement` parsing and node, for performance reasons.

We will then not conform to the parsing and evaluation described in the ES, which possibly introduces errors and makes it harder to follow.

Razican commented on boa/src/syntax/parser/mod.rs:
```diff
@@ -139,6 +139,9 @@ where
 }
 
+/// The possible TokenKind which indicate the end of a case statement.
+const SCRIPT_BREAK_TOKENS: [TokenKind; 1] = [TokenKind::Punctuator(Punctuator::CloseBlock)];
```
Is this defined somewhere? I believe a script will never end with a `}`, right? I think this list should be empty, and in that case, we should add the exception to the parser so that it doesn't fail with an `AbruptEnd` error.

AngelOnFira replied:
No, I just added it cause I thought I needed to. I'm going to reassess each one and see what it's proper closing should be.

AngelOnFira replied:
Ok, I think this should be resolved. I'm not sure I understand what an exception should be, I ran the tests and I'm getting the same percentage. Can you give me an example of where this might occur?

Razican commented on boa/src/syntax/parser/statement/mod.rs:
```diff
@@ -214,22 +214,24 @@ where
         allow_return: R,
         break_when_closingbraces: bool,
         in_block: bool,
+        break_nodes: &'a [TokenKind],
```
I think it makes sense to enforce this to be `&'static`, and that way we remove some extra lifetimes. I don't think it makes sense to customize this on runtime.

AngelOnFira replied:
This is purely due to my lack of understanding Rust lifetimes past what the compiler tells me to do :sweat_smile: I'll switch it up :100:

Razican reviewed (commented):
I like this new implementation, should solve the issue we are having with empty statements :)

Razican commented:
After some discussion on discord, it seems that EmptyStatements are indeed important, and they will be implemented fully in the future, as @croraf mentions. For now, I would merge this as soon as the review comments are solved.

croraf commented:
Can we merge just the refactor part (which is the majority of this PR), without the semicolon part?

AngelOnFira commented:
Sorry @croraf I'll be working through the comments today, just got hit with a ton of work this week :sweat_smile:

croraf commented:
@AngelOnFira No hurry, we were just discussing.

AngelOnFira commented:
Ok @croraf I've reverted the semicolon changes for this PR. @Razican I believe I've addressed all of your comments, but a few need a bit more discussion to confirm that they are done. Please let me know if there is anything else to take care of! :smile:

croraf commented:
Cool. After this bigger addressing you can make a new PR for the semicolon changes if you wish, it should be very easy to solve that.

AngelOnFira commented:
Oops still have to make the tests :P

AngelOnFira commented:
Actually I think the tests don't apply here since this is just a refactor.

AngelOnFira commented:
@Razican @Lan2u Just want to check in. I think this PR is ready to go, but please ping me if anything else needs to be done :smile:

Razican reviewed (approved):
Looks good, thanks!

### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 78,415 | 78,415 | 0 |
  | Passed | 18,945 | 18,945 | 0 |
  | Ignored | 15,547 | 15,547 | 0 |
  | Failed | 43,923 | 43,923 | 0 |
  | Panics | 1,127 | 1,127 | 0 |
  | Conformance | 24.16 | 24.16 | 0.00% |

croraf commented:
@AngelOnFira Gj. Are you willing to make another PR with the semicolon issue fix?

AngelOnFira commented:
@croraf ya, I'll try to do that now that this is merged.
